packages/neon-core/src/sc/ScriptBuilder.ts

```diff
@@ -31,121 +38,109 @@ function isValidValue(value: any): boolean {
   return false;
 }
 
-/**
- * Retrieves a single AppCall from a ScriptBuilder object.
- * Returns ScriptIntents starting from the beginning of the script.
- * This is based off the pointer in the stream.
- * @param sb
- * @returns A single ScriptIntent if available.
- */
+// TODO
 function retrieveAppCall(sb: ScriptBuilder): ScriptIntent | null {
-  const output: ScriptIntent = {
-    scriptHash: "",
-    args: []
-  };
-
-  while (!sb.isEmpty()) {
-    const b = sb.read();
-    const n = parseInt(b, 16);
-    switch (true) {
-      case n === 0:
-        output.args!.unshift(0);
-        break;
-      case n < 75:
-        output.args!.unshift(sb.read(n));
-        break;
-      case n >= 81 && n <= 96:
-        output.args!.unshift(n - 80);
-        break;
-      case n === 193:
-        const len = output.args!.shift();
-        const cache = [];
-        for (let i = 0; i < len; i++) {
-          cache.unshift(output.args!.shift());
-        }
-        output.args!.unshift(cache);
-        break;
-      case n === 102:
-        sb.pter = sb.str.length;
-        break;
-      case n === 103:
-        output.scriptHash = reverseHex(sb.read(20));
-        output.useTailCall = false;
-        return output;
-      case n === 105:
-        output.scriptHash = reverseHex(sb.read(20));
-        output.useTailCall = true;
-        return output;
-      case n === 241:
-        break;
-      default:
-        throw new Error(`Encounter unknown byte: ${b}`);
-    }
-  }
-  if (output.scriptHash === "") {
-    return null;
-  }
-  return output;
+  throw new Error("Not Implemented");
 }
 
 /**
  * Builds a VM script in hexstring. Used for constructing smart contract method calls.
  * @extends StringStream
  */
 export class ScriptBuilder extends StringStream {
+  public fee: Fixed8;
+
+  public constructor(str = "") {
+    super(str);
+    this.fee = new Fixed8(0);

```

Comment:
Why will this result in a large number of errors?
Developer will not have a headache, he can even overlook it because fees counting will be done silently and synchronously while he constructs a script.

Reply:
```
var sb = new ScriptBuilder("abcd");
```

We cannot initialize the ScriptBuilder with anything or the fees will be wrong. We also cannot use the ScriptBuilder to reverse scripts. We can do the fees accounting at the end with a `build` call instead of making sure every emit method adjusts the fee correctly.

Reply:
> We cannot initialize the ScriptBuilder with anything or the fees will be wrong. We also cannot use the ScriptBuilder to reverse scripts. 

You can do all the things, but you've got to realize that the counting of fees will not be right.
> We can do the fees accounting at the end with a build call instead of making sure every emit method adjusts the fee correctly.

I considered about this before. 2 problems:
 - This needs a decoding logic of script, the logic may be complex and expensive
 - This require more process and time-costing because a surplus reverse engineering will be done.

Reply:
dont worry abt cpu cycles at the moment decoding scripts should be so much cheaper than all the crypto stuff we are doing.

Effectively, you are providing a feature correctly which only works with empty string. Im sure there are more ways to do this.

Reply:
@snowypowers I check your design for `ScriptBuilder`, now I realize that this class is not only to build a script, but also to read a script(that's why it extends StringStream, right?).
Considering this, fee calculation is not proper to be included in this class.
And BTW, should we consider redesigning ScriptBuilder to  split the builder and reader? So far this name is not proper enough.
Anyway I have removed fee calculation. Let's see how to improve this.

Reply:
yeah i think we should split the builder and reader.